memory/unmergeable_index.go

```diff
@@ -0,0 +1,177 @@
+	DB         string // required for engine tests with driver
+	DriverName string // required for engine tests with driver
+	Tbl        *Table // required for engine tests with driver
+	TableName  string
+	Exprs      []sql.Expression
+}
+
+func (u *UnmergeableDummyIndex) Database() string { return u.DB }
+func (u *UnmergeableDummyIndex) Driver() string   { return u.DriverName }
+
+func (u *UnmergeableDummyIndex) Expressions() []string {
+	var exprs []string
+	for _, e := range u.Exprs {
+		exprs = append(exprs, e.String())
+	}
+	return exprs
+}
+
+func (u *UnmergeableDummyIndex) Get(key ...interface{}) (sql.IndexLookup, error) {
+	return &UnmergeableIndexLookup{
+		key: key,
+		idx: u,
+	}, nil
+}
+
+// UnmergeableIndexLookup is the only IndexLookup in this package that doesn't implement Mergeable, and therefore
+// can't be merged with other lookups.
+type UnmergeableIndexLookup struct {
+	key []interface{}
+	idx *UnmergeableDummyIndex
+}
+
+// dummyIndexValueIter does a very simple and verifiable iteration over the table values for a given index. It does this
+// by iterating over all the table rows for a partition and evaluating each of them for inclusion in the index. This is
+// not an efficient way to store an index, and is only suitable for testing the correctness of index code in the engine.
+type dummyIndexValueIter struct {
+	tbl *Table
+	partition sql.Partition
+	matchExpression sql.Expression
+	values [][]byte
+	i int
+}
+
+func (u *dummyIndexValueIter) Next() ([]byte, error) {
+	err := u.initValues()
+	if err != nil {
+		return nil, err
+	}
+
+	if u.i < len(u.values) {
+		valBytes := u.values[u.i]
+		u.i++
+		return valBytes, nil
+	}
+
+	return nil, io.EOF
+}
+
+func (u *dummyIndexValueIter) initValues() error {
+	if u.values == nil {
+		rows, ok := u.tbl.partitions[string(u.partition.Key())]
+		if !ok {
+			return fmt.Errorf(
+				"partition not found: %q", u.partition.Key(),
+			)
+		}
+
+		for i, row := range rows {
+			ok, err := sql.EvaluateCondition(sql.NewEmptyContext(), u.matchExpression, row)
+			if err != nil {
+				return err
+			}
+
+			if ok {
+				encoded, err := encodeIndexValue(&indexValue{
+					Pos: i,
+				})
+
+				if err != nil {
+					return err
+				}
+
+				u.values = append(u.values, encoded)
+			}
+		}
+	}
+
+	return nil
+}
+
+func getType(val interface{}) (interface{}, sql.Type) {
+	switch val := val.(type) {
+	case int8:
+		return int64(val), sql.Int64
+	case uint8:
+		return int64(val), sql.Int64
+	case int16:
+		return int64(val), sql.Int64
+	case uint16:
+		return int64(val), sql.Int64
+	case int32:
+		return int64(val), sql.Int64
+	case uint32:
+		return int64(val), sql.Int64
+	case int64:
+		return int64(val), sql.Int64
+	case uint64:
+		return int64(val), sql.Int64
+	case float32:
+		return float64(val), sql.Float64
+	case float64:
+		return float64(val), sql.Float64
+	case string:
+		return val, sql.Text
+	default:panic(fmt.Sprintf("Unsupported type for %v of type %T", val, val))
```

Comment:
Hard to see from GitHub where this is all used and the impact it would have, but this is converting everything to 64-bit precision and returning the associated SQL type. I think it will but I wanna make sure: this will only be used from the index driver right? As in any value passed in won't make it's way into an insertion?

I ask in the case of a column expecting an `int16` but receiving an `int64` for storage. There may be a benefit to working with specific types here that I'm not seeing since it's hard to gather the bigger picture from here (like that comparing two ints of different sizes thing you ran into before), but I wanna check.

Reply:
Yeah, this is just for comparison purposes for SELECT statements. Using the widest possible numeric type makes the comparison more straightforward.

Reply:
Okay, just wanted to double check!